packages/xo-server/src/xo-mixins/backups-ng/index.js

```diff
@@ -1032,110 +1042,141 @@ export default class BackupNg {
+
+      const bypassVdiChainsCheck: boolean = getSetting(
+        settings,
+        'bypassVdiChainsCheck',
+        [vmUuid, '']
+      )
+      if (!bypassVdiChainsCheck) {
+        xapi._assertHealthyVdiChains(vm)
+      }
 
-    const bypassVdiChainsCheck: boolean = getSetting(
-      settings,
-      'bypassVdiChainsCheck',
-      [vmUuid, '']
-    )
-    if (!bypassVdiChainsCheck) {
-      xapi._assertHealthyVdiChains(vm)
-    }
+      const offlineSnapshot: boolean = getSetting(settings, 'offlineSnapshot', [
+        vmUuid,
+        '',
+      ])
+      const startAfterSnapshot = offlineSnapshot && vm.power_state === 'Running'
+      if (startAfterSnapshot) {
+        await wrapTask(
+          {
+            logger,
+            message: 'shutdown VM',
+            parentId: taskId,
+          },
+          xapi.shutdownVm(vm)
+        )
+      }
 
-    const offlineSnapshot: boolean = getSetting(settings, 'offlineSnapshot', [
-      vmUuid,
-      '',
-    ])
-    const startAfterSnapshot = offlineSnapshot && vm.power_state === 'Running'
-    if (startAfterSnapshot) {
-      await wrapTask(
+      exported = (await wrapTask(
         {
           logger,
-          message: 'shutdown VM',
+          message: 'snapshot',
           parentId: taskId,
+          result: _ => _.uuid,
         },
-        xapi.shutdownVm(vm)
-      )
-    }
+        xapi._snapshotVm(
+          $cancelToken,
+          vm,
+          `[XO Backup ${job.name}] ${vm.name_label}`
+        )
+      ): any)
 
-    let snapshot: Vm = (await wrapTask(
-      {
-        logger,
-        message: 'snapshot',
-        parentId: taskId,
-        result: _ => _.uuid,
-      },
-      xapi._snapshotVm(
-        $cancelToken,
-        vm,
-        `[XO Backup ${job.name}] ${vm.name_label}`
-      )
-    ): any)
+      if (startAfterSnapshot) {
+        ignoreErrors.call(xapi.startVm(vm))
+      }
 
-    if (startAfterSnapshot) {
-      ignoreErrors.call(xapi.startVm(vm))
-    }
+      await wrapTask(
+        {
+          logger,
+          message: 'add metadata to snapshot',
+          parentId: taskId,
+        },
+        exported.update_other_config({
+          'xo:backup:datetime': exported.snapshot_time,
+          'xo:backup:job': jobId,
+          'xo:backup:schedule': scheduleId,
+          'xo:backup:vm': vmUuid,
+        })
+      )
 
-    await wrapTask(
-      {
-        logger,
-        message: 'add metadata to snapshot',
-        parentId: taskId,
-      },
-      snapshot.update_other_config({
-        'xo:backup:datetime': snapshot.snapshot_time,
-        'xo:backup:job': jobId,
-        'xo:backup:schedule': scheduleId,
-        'xo:backup:vm': vmUuid,
-      })
-    )
+      exported = await xapi.barrier(exported.$ref)
 
-    snapshot = await xapi.barrier(snapshot.$ref)
+      if (mode === 'delta') {
+        baseSnapshot = findLast(
+          snapshots,
+          _ => 'xo:backup:exported' in _.other_config
+        )
 
-    let baseSnapshot
```

Comment:
Why did you move this?

Reply:
There have a mutualized code between full backups and delta backups which is linked to snapshot and its retention. So, to not change the actual implementation of snapshot deletion because it should not be in case of the offline backup mode, i decided to move this peace of code before the backup step.